File: src/Material.cc

#include <string>

#include "Material.h"

namespace HeatFlow {

Material::Material() 
{
    this->name_ = "";
	this->conductivity_ = 0.591; // water
	this->density_ = 1.0; // water at 4-5 degrees C
}
Material::~Material()
{

}

// I/O helpers
int Material::to_xml(int index, std::string& xml_buffer)
{
    // <material index="%i"><name>%s</name></material> -> 46 characters for the markup
    std::string indexStr = std::to_string(index);
	std::stringstream ss;

	ss << "<Material id=\"" << index << "\"><name>"
		<< this->name_ << "</name><density>"
		<< this->density_ << "</density><conductivity>"
		<< this->conductivity_ << "</conductivity></Material>";

    
	xml_buffer.assign(ss.str());
    return 1;
}

} // namespace HeatFlow

File: lib/material.rb
class Material
    attr_accessor :name, :density, :conductivity, :index, :specific_heat

    def initialize(specific_heat: 1.0, name: "New Material", density: 1, conductivity: 1, index: -1)
        @name = name
        @density = density
        @conductivity = conductivity
        @index = index
        @specific_heat = specific_heat
    end
    
    def to_h
        {
            "name" => @name,
            "density" => @density,
            "conductivity" => @conductivity,
            "index" => @index,
            "specific_heat" => @specific_heat
        }
    end

    def from_h(config_hash)
        @name = config_hash["name"]
        @density = config_hash["density"]
        @conductivity = config_hash["conductivity"]
        @index = config_hash["index"].to_i
        @specific_heat = config_hash['specific_heat'].to_f
    end
    
    def to_s_pretty
        return to_h.to_s_pretty
    end
end
